Check if plugins file/folder exists before load/require

Check if a plugin file/folder exsits before to require/load.
Backends crash if file is not found:

```sh
$ npm run start

> backend@1.0.0 start
> ./node_modules/.bin/nodemon index.js

[nodemon] 2.0.12
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): *.*
[nodemon] watching extensions: js,mjs,json
[nodemon] starting `node index.js`
Starting OpenHaus v1.0.0...
Error in plugin: Error: Cannot find module '/home/marc/projects/OpenHaus/backend/plugins/d11e7f38-91cb-4f32-89e8-3452f624bb47/index.js'
Require stack:
- /home/marc/projects/OpenHaus/backend/components/plugins/class.plugin.js
- /home/marc/projects/OpenHaus/backend/components/plugins/index.js
- /home/marc/projects/OpenHaus/backend/index.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:962:15)
    at Function.Module._load (internal/modules/cjs/loader.js:838:27)
    at Module.require (internal/modules/cjs/loader.js:1022:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Plugin.boot (/home/marc/projects/OpenHaus/backend/components/plugins/class.plugin.js:38:30)
    at /home/marc/projects/OpenHaus/backend/index.js:361:20
    at Array.forEach (<anonymous>)
    at /home/marc/projects/OpenHaus/backend/index.js:356:8
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/marc/projects/OpenHaus/backend/components/plugins/class.plugin.js',
    '/home/marc/projects/OpenHaus/backend/components/plugins/index.js',
    '/home/marc/projects/OpenHaus/backend/index.js'
  ]
}

```

--- a/docs/README.md
+++ b/docs/README.md
@@ -32,6 +32,13 @@ The collection can be easy imported and is read-to-use.
 | CORS_ORIGIN              | `*`                             | Where does the request come from: https://developer.mozilla.org/en/docs/Web/HTTP/Headers/Access-Control-Allow-Origin                   |
 | CORS_HEADERS             | `X-AUTH-TOKEN, *`               | What custom headers is the client allowed to send: https://developer.mozilla.org/en/docs/Web/HTTP/Headers/Access-Control-Allow-Headers |
 | CORS_METHODS             | `GET, PUT, PATCH, DELETE, POST` | What HTTP Methods are allowed: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Methods                  |
+| VAULT_MASTER_PASSWORD    |                                 | Vault component master Password, need to be set to start OpenHaus                                                                      |
+| VAULT_BLOCK_CIPHER       | `aes-256-cbc`                   | Vault encryption method                                                                                                                |
+| VAULT_AUTH_TAG_BYTE_LEN  | `16`                            | Vault auth tag length in bytes                                                                                                         |
+| VAULT_IV_BYTE_LEN        | `16`                            | Vault "initial vector" value size in bytes                                                                                             |
+| VAULT_KEY_BYTE_LEN       | `32`                            | Vault key size in bytes                                                                                                                |
+| VAULT_SALT_BYTE_LEN      | `16`                            | Vault salt size in bytes                                                                                                               |
+
 
 ## Components
 - [devices](./components/devices.md)
@@ -51,6 +58,7 @@ The collection can be easy imported and is read-to-use.
 - [promisify(worker, cb)](./helper.md#promisifyworker-cb)
 - [queue(counter, cb)](./helper.md#queuecounter-cb)
 - [timeout(time, cb)](./helper.md#timeouttime-cb)
+- [infinity(worker, delay)](./helper.md#infinityworker-cb)
 
 ## System
 - [components](./system/components.md)

--- a/docs/README.md
+++ b/docs/README.md
@@ -58,7 +58,7 @@ The collection can be easy imported and is read-to-use.
 - [promisify(worker, cb)](./helper.md#promisifyworker-cb)
 - [queue(counter, cb)](./helper.md#queuecounter-cb)
 - [timeout(time, cb)](./helper.md#timeouttime-cb)
-- [infinity(worker, delay)](./helper.md#infinityworker-cb)
+- [infinity(worker, delay)](./helper.md#infintyworker-delay)
 
 ## System
 - [components](./system/components.md)

--- a/docs/components/vault.md
+++ b/docs/components/vault.md
@@ -1,0 +1,127 @@
+# [🗀](../../components/vault) VAULT
+
+> [C_VAULT](./) extends [COMMON_COMPONENT](../system/component.md)
+
+Component to handle username/password & API tokens which are used to authenticate gains servcies & devices.
+
+
+---
+
+
+## Table of contents 
+- Schema
+- Methods
+  - [Add](#adddatacb) *Add new deivce*
+  - [Get](#getidcb) *Get existing Vault*
+  - [Remove](#removeidcb) *Removes exisiting Vault*
+  - [Update](#updateid-datacb) *Update Vault*
+  - [Find](#findobjcb) *Find Vault in database*
+- Classes
+  - [class.secret.js]() *Represents a single field instance*
+
+****
+---
+
+
+## Schema
+| Name                  | Type   | Required | Default value | Description                                                                      |
+| --------------------- | ------ | -------- | ------------- | -------------------------------------------------------------------------------- |
+| _id                   | String |          | [ObjectId]()  | MongoDB ObjectID converted to String                                             |
+| name                  | String | x        |               | Human friendly name e.g. "Weather API", "ZigBee Gateway"                         |
+| identifier            | String | x        |               | Identifier for Plugin/Hardcoded                                                  |
+| fields                | Array  |          | `[]`          | Array fof fields needed, e.g. "Username", "Password", "API Token"                |
+| fields[0]._id         | String |          | [ObjectId]()  | MongoDB ObjectID converted to string                                             |
+| fields[0].name        | String | x        |               | Humany friendly name                                                             |
+| fields[0].description | String |          | `null`        | Field description                                                                |
+| fields[0].key         | String | x        |               | Hardcoded value in Plugin                                                        |
+| fields[0].value       | String |          |               | Encrypted field value                                                            |
+| timestamps            | Object |          |               | Timestamps, when something was done to the object e.g. `created`/`updated`       |
+| timestamps.created    | Number |          | `Date.now()`  | Unix timestamp, set when [`.add`](#adddatacb) is called to `Date.now()`          |
+| timestamps.updated    | Number |          | `null`        | Unix timestamp, set to `Date.now()` when [`.update`](#updateid-datacb) is called |
+
+### Example
+
+```json
+{
+    "_id": "6199783105707e5faa429368",
+    "name": "ZigBee Gateway",
+    "identifier": "ZIGBEE_GW_RASPBEE",
+    "fields": [
+        {
+            "name": "API Token",
+            "key": "API_TOKEN",
+            "_id": "6199783105707e5faa429369",
+            "description": null,
+            "value": "128ab9c6adb33c19878c0c67770618d3:74e4314d1083cfa5d9cc2550d97d1a6a:3de9d2c47a4e8e38e0886870d47c729ee994ed0fa4e24ba906180a954732d5a081a6ca57c08db289f51097c42f0fb992708fef34c561a92a2341bfd30763f2fd"
+        },
+        {
+            "name": "Username",
+            "key": "USERNAME",
+            "description": null,
+            "value": "32d5ab650c265d455e75df6fbf7678a9:59fcc731e92aef826fa1e13498666816:e63217c7d1e86806d6986fa3cb3da731ea95e1383cfb524895e48db5c963fc25",
+            "_id": "619e8169bf178d249985aab6"
+        },
+        {
+            "name": "User Password",
+            "key": "PASSWORD",
+            "description": null,
+            "value": "3351583b3a22daaaf283b2cf37d2fde1:430214393d16835aeb0ab8deb0128fd1:2fe7c9d0c7841ab0bcd126cd25c86806",
+            "_id": "619e815abf178d249985aab5"
+        }
+    ],
+    "timestamps": {
+        "created": 1637447729869,
+        "updated": 1637783734049
+    }
+}
+```
+
+
+---
+
+
+## Methods
+
+### __add(data[,cb])__
+* `data` {Object} Object structure as defined in schema.
+* `cb` {Function} Callback function, if absent, a promise is returned.
+> Add a new Vault to the database and creates a Vault instance
+
+### __get(id[,cb])__
+* `id` {String} Vault id as defined in schema.
+* `cb` {Function} Callback function, if absent, a promise is returned.
+> Get a Vault from the `.items` array
+
+### __remove(id[,cb])__
+* `id` {String} Vault id as defined in schema.
+* `cb` {Function} Callback function, if absent, a promise is returned.
+> Removes a Vault from the database
+
+### __update(id, data[,cb])__
+* `id` {String} Vault id as defined in schema.
+* `data` {Object} Properties to update on Room `id` object/instance.
+* `cb` {Function} Callback function, if absent, a promise is returned.
+> Updates a Vault object in database & Vault instance in `.items` array
+
+### __find(obj[,cb])__
+* `obj` {Object} Key/value to look for Vault.
+* `cb` {Function} Callback function, if absent, a promise is returned.
+> Find Vault with mathing key/values pair in the `.items` array
+
+### __encrypt(id, data[,cb])__
+* `id` {String} Vault id (ObjectID)
+* `data` {Object} Field key/value object to encrypt values
+* `cb` {function} Callback function, if absent, a promise is returned.
+> Encrypt data values & update document
+
+### __decrypt(id[,cb])__
+* `id` {String} Vault id (ObjectID)
+* `cb` {function} Callback function, if absent, a promise is returned.
+> Decrypt field values
+
+---
+
+## Classes
+
+### class.secret.js
+> Represent a single field instance

--- a/docs/helper.md
+++ b/docs/helper.md
@@ -80,3 +80,12 @@ Returns a function, that needs to be called *x* times.
 >  Calls a callback function  automaticly after `time`, unless the returned function is not called before timeout is reached.
 
 Return a function that can be called to trigger `cb`.
+
+
+### __infinty(worker, delay)__
+* `worker` {Function} Worker callback that is with a delay of `time` called.
+* `delay` {Number} Delay to wait before calling`worker` again.
+> Calls the worker function, passes a "retry" function as only argument.
+> If the retry function is called, its wait the time set as `delay` and call then the worker code again
+
+Ifinity function handling. For "redo" every time the "retery" function is called.

--- a/docs/helper.md
+++ b/docs/helper.md
@@ -12,6 +12,7 @@ This collection should help you to avoid the need of external modules like "unde
 - [promisify(worker, cb)](#promisifyworker-cb)
 - [queue(counter, cb)](#queuecounter-cb)
 - [timeout(time, cb)](#timeouttime-cb)
+- [infinty(worker, cb)](#infintyworker-delay)
 
 
 ### __debounce(func, wait[, immediate])__

--- a/routes/index.js
+++ b/routes/index.js
@@ -6,6 +6,7 @@ const C_USERS = require("../components/users");
 const C_ROOMS = require("../components/rooms");
 const C_DEVICES = require("../components/devices");
 const C_ENDPOINTS = require("../components/endpoints");
+const C_VAULT = require("../components/vault");
 //const C_SCENES = require("../components/scenes");
 
 const { encode } = require("../helper/sanitize");
@@ -80,6 +81,7 @@ module.exports = (server) => {
         const roomsRouter = express.Router();
         const devicesRouter = express.Router();
         const endpointsRouter = express.Router();
+        const vaultRouter = express.Router();
         //const scenesRouter = express.Router();
         const eventsRouter = express.Router();
 
@@ -128,6 +130,11 @@ module.exports = (server) => {
         require("./rest-handler.js")(C_ENDPOINTS, endpointsRouter);
         require("./router.api.endpoints.js")(app, endpointsRouter);
 
+        // http://127.0.0.1/api/vaults
+        api.use("/vault", vaultRouter);
+        require("./rest-handler.js")(C_VAULT, vaultRouter);
+        require("./router.api.vault.js")(app, vaultRouter);
+
         // http://127.0.0.1/api/scenes
         //api.use("/scenes", scenesRouter);
         //require("./rest-handler.js")(C_SCENES, scenesRouter);
@@ -146,14 +153,14 @@ module.exports = (server) => {
         });
 
         // https://expressjs.com/de/guide/error-handling.html
-        app.use((req, res, next, error) => {
+        app.use((err, req, res) => {
 
-            console.error(error.stack);
+            console.error(err.stack, req);
 
             res.status(500);
 
             if (process.env.NODE_ENV !== "production") {
-                res.end(error.message);
+                res.end(err.message);
             } else {
                 res.end();
             }

--- a/routes/router.api.vault.js
+++ b/routes/router.api.vault.js
@@ -1,0 +1,33 @@
+const C_VAULT = require("../components/vault");
+
+module.exports = (app, router) => {
+
+    router.post("/:_id/encrypt", (req, res) => {
+        C_VAULT.encrypt(req.item._id, req.body, (err) => {
+            if (err) {
+
+                res.status(400).end(err);
+
+            } else {
+
+                res.json(req.item);
+
+            }
+        });
+    });
+
+    router.post("/:_id/decrypt", (req, res) => {
+        req.item.decrypt((err, values) => {
+            if (err) {
+
+                res.status(400).end(err);
+
+            } else {
+
+                res.json(values);
+
+            }
+        });
+    });
+
+};

